ReactAndroid/src/main/java/com/facebook/react/modules/settings/SettingsModule.java

```diff
@@ -0,0 +1,180 @@
+  static final String TAG = "SettingsMmodule";
+
+  private static boolean sSettingsModuleInstantiated = false;
+
+  static private String sFilename = "";
+  private Boolean ignoringUpdates = false;
+  private SharedPreferences mPreferences;
+
+  public SettingsModule(ReactApplicationContext context) {
+    super(context);
+
+  }
+
+  @Override
+  public String getName() {
+    return SettingsModule.NAME;
+  }
+
+  private SharedPreferences ensurePreferences() {
+    if (mPreferences == null) {
+      if (SettingsModule.sFilename.length() > 0) {
+        mPreferences = getReactApplicationContext().getSharedPreferences(SettingsModule.sFilename, Context.MODE_PRIVATE);
+      } else if (getCurrentActivity() != null) {
+        mPreferences = getCurrentActivity().getPreferences(Context.MODE_PRIVATE);
+      } else {
+        mPreferences = getReactApplicationContext().getSharedPreferences(DEFAULT_FILE_NAME, Context.MODE_PRIVATE);
+      }
+      sSettingsModuleInstantiated = true;
+      mPreferences.registerOnSharedPreferenceChangeListener(this);
+    }
+
+    return mPreferences;
+  }
+
+  /***
+   * Configure the Settings Module to use a SharedPreferences file with a specific name as its backing store.
+   * Useful if the Settings Module needs to use an existing SharedPreferences file or to ensure that Settings Modules
+   * belonging to different ReactPackages will all share the same SharedPreferences file for their backing store.
+   * <p></p>
+   * <b>Note:</b>this method may only be called before a Settings Module initialization (i.e. before a ReactMainPackage is instantiated)
+   * An app would typically invoke this function in the app's main activity before setting up the React Native environment
+   * @param filename the name of the SharedPreferences file to use or create for the backing store
+   * @throws IllegalStateException if called after instantiation of a Settings module
+   */
+  static public void setSharedPreferencesFilename(String filename) {
+    if (sSettingsModuleInstantiated) {
+      throw new IllegalStateException("Attempt to set Settings Module SharedPreferences file name after a Settings Module instantiation");
+    }
+    sFilename = filename;
+  }
+
+  @Override
+  public void onCatalystInstanceDestroy() {
+    ensurePreferences().unregisterOnSharedPreferenceChangeListener(this);
+  }
+
+  @ReactMethod
+  public void setValues(ReadableMap values) {
+    this.ignoringUpdates = true;
+
+    ReadableMapKeySetIterator iterator = values.keySetIterator();
+    SharedPreferences.Editor editor = mPreferences.edit();
+
+    while (iterator.hasNextKey()) {
+      String key = iterator.nextKey();
+      ReadableType type = values.getType(key);
+
+      switch (type) {
+        case Null:
+          editor.remove(key);
+          break;
+        case Boolean:
+          editor.putBoolean(key, values.getBoolean(key));
+          break;
+        case Number:
+          editor.putFloat(key, (float) values.getDouble(key));
```

Comment:
I am aware that the settings values live in javascript and are also asynchronously (although I would't say lazily) sent to native land and persisted.

The problem here is that you can store an object to preferences, for example `{name: "John", age: null }` and when you read the preferences somewhere else, you the module returns the same what you have stored (because the value is read from JS). But then you restart the app and what you get is `{name: "John"}` - the `age` field is lost!

With the current implementation, people can store things without knowing they weren't saved or that they weren't saved exactly the way they were meant to be saved (e.g. because the settings module does not support storing nulls or because there was a trouble casting their large number they wanted to save). This silent failing is very developer-unfriendly. That is why I suggested that a promise would be helpful - if such case happens, the promise could reject.

I think it pays off to do things right here - many people will likely use this module (myself included, which is why I did the review). Now, I am not a person that has the rights to merge this PR, so maybe you can go ahead and ask some core contributor to merge, but I think they will also want to make the module more error-proof. I think I have said all that I wanted, so good luck getting this merged 👍